File: client/core/ClientMessageArray.js
define(function() {

let module = {
    ACK_TIMEOUT: 10000,
};

module.ClientMessageArray = class {
    constructor(flazh, ms=[]) {
        ms = ms.concat();
        this.flazh = flazh;
        this.messages = [];

        this.sent = false;
        this.processedByServer = false;
        this.timedOut = false;

        for(let m of ms) this.pushMessage(m);
    }

    pushMessage(m) {
        let fzh = this.flazh;

        if(this.sent) throw new Error('cannot add a message to an array that has already been sent');

        if(m.type === undefined) throw new Error('a client message is missing the type attribute');
        if(fzh.MSG_TYPES_CLI.indexOf(m.type) < 0) throw new Error('unsupported client message type: ' + m.type);

        this.messages.push(m);
    }

    onServerAck(kwarg={}) {
        kwarg = Object.assign({
            ack: false,
            data: undefined,
            timedOut: false,
        }, kwarg);
        
        let fzh = this.flazh;

        if(kwarg.ack) {
            if(kwarg.timedOut) throw new Error('ack & timedOut cannot be used together');
            this.processedByServer = true;
            console.log('server has processed messages: ', kwarg.data);
        } else {
            this.processedByServer = false;
            if(kwarg.timedOut) {
                console.log('server acknowledgement timed out');
                this.timedOut = true;
            } else {
                console.log('an error occurred on the server while processing messages');
            }
        }

        fzh.onMessageArrayOutcome(this);
    }

    send() {
        let thisObj = this;
        let fzh = this.flazh;
        let sock = fzh.sock;
        let ms = this.messages;

        if(this.sent) throw new Error('the client message array has already been sent');

        if(ms.length) {
            sock.emit('client_message_array', ms, (data) => {
                thisObj.onServerAck({'ack': true, 'data': data});
            });

            window.setTimeout((data) => {
                thisObj.onServerAck({timedOut: true});
            }, module.ACK_TIMEOUT);
        }

        this.sent = true;

        console.log(ms.length + ' messages sent to the server');
    }
};

return module;});

File: client/ClientMessageArray.py
from core.strings import genRandomStr
from core import ConcurSensitiveObjs

ACK_TIMEOUT_SEC = 10
DFLT_SEC_TO_LIVE = 60

class ClientMessageArray:
    def __init__(self, fzhc, ms=()):
        self.ref = genRandomStr()

        ms = list(ms)
        self.flazhClient = fzhc

        if not fzhc.lock: raise Exception(f'flachClient lock is undefined')
        self.concur = concur = ConcurSensitiveObjs(fzhc.lock)

        with concur:
            concur.messages = []
            concur.sent = False
            concur.processedByServer = False
            concur.timedOut = False
            concur.expired = False
            concur.secToLive = DFLT_SEC_TO_LIVE
            concur.callbacks = {}

            for m in ms:
                self.pushMessages(m)

    def pushMessage(self, m):
        """If m.responseValidForSec is given
        (and is greater than other messages', or the default value)
        it sets for how many seconds to keep this ClientMessageArray after
        it is sent to the server.
        
        If a function is provided as m.callback it will be kept in this
        object until it times out/expires, so that the server can trigger
        it, if necessary.
        """

        fzhc = self.fzhc
        concur = self.concur

        with concur:
            if self.sent: raise Exception('cannot add a message to an array that has already been sent')

            typ = m.get('type')
            if typ is None: raise Exception(f'a client message is missing the type attribute')

            if typ not in fzhc.concur.MSG_TYPES_CLI: raise Exception(f'unsupported client message type: {typ}')

            m = dict(m)
            m.clientMessageArray = self.ref

            vsec = m.get('responseValidForSec')
            if vsec and vsec > concur.secToLive:
                concur.secToLive = vsec

            cb = m.get('callback')
            if cb:
                cbk = genRandomStr()
                concur.callbacks[cbk] = cb
                m['callback'] = cbk

            self.messages.push(m)

    def onStatusChange(ack=False, data=None, timedOut=False, expired=False):
        fzhc = self.fzhc
        # TODO complete
